### ISSUE TITLE
Error trying to connect to local Virtuoso instance running via Docker on Fedora

### ISSUE THREAD

**Issue by eltonfss** [OpenLinkSoftware/Virtuoso-SQLAlchemy#3]
# Issue Description

When trying to connect to connect to Virtuoso via SQLAlchemy I get the error `[IM002] [unixODBC][Driver Manager]Data source name not found and no default driver specified (0) (SQLDriverConnect)`.
I've tried multiple workarounds to configure the driver properly on Fedora, but have not been able to get over this particular error.

The test consists of initiating Virtuoso via Docker Compose and running a python script equivalent to the following:
```python
from sqlalchemy import create_engine

metadata = MetaData()
...
# create multiple tables associated to metadata
...

engine = create_engine("virtuoso+pyodbc://dba:dba@VOS")

with engine.connect().execution_options(isolation_level="AUTOCOMMIT") as conn:
    with engine.begin():
        metadata.drop_all(conn, checkfirst=True)
        metadata.create_all(conn, checkfirst=False)
    conn.commit()
```

The error is being raised the line `with engine.connect().execution_options(isolation_level="AUTOCOMMIT") as conn:`, which indicates that is unlikely associated to the specific tables I'm trying to create.

Also, after searching this error I found this [StackOverflow post](https://stackoverflow.com/questions/16280304/pyodbc-data-source-name-not-found-and-no-default-driver-specified) whose answer indicates that the issue is likely related to some problem with the ODBC driver installation/configuration. 

Would you be able to provide further detail on how the driver could be installed/configured on Fedora and how I might solve this particular issue?

## End of Error Stack Trace
```
... lib64/python3.11/site-packages/sqlalchemy/engine/default.py", line 616, in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sqlalchemy.exc.InterfaceError: (pyodbc.InterfaceError) ('IM002', '[IM002] [unixODBC][Driver Manager]Data source name not found and no default driver specified (0) (SQLDriverConnect)')
(Background on this error at: https://sqlalche.me/e/20/rvf5)
```

### Virtual Machine (VM) Setup

```
VM OS: Fedora release 39 (Thirty Nine)
VM CPU / RAM / Disk: 2 / 4GB / 32 GB
Hypervisor: Oracle VirtualBox 7.0.4
Host OS: MacOS 14.2.1 (Sonoma)
```

### Python Setup
```
Python 3.11.7
pip 24.0 from /home/eltons/virtualenvs/dbms-benchmarking-framework/lib64/python3.11/site-packages/pip (python 3.11)

# dependencies
virtuoso-sqlalchemy @ git+https://github.com/OpenLinkSoftware/Virtuoso-SQLAlchemy.git@1f88bad4142dfbb030934be3a6fc87d9b6374529
SQLAlchemy==2.0.25
pyodbc==5.1.0
```

### docker-compose.yaml
```yaml
version: '3.1'

services:

  db:
    image: openlink/virtuoso-opensource-7:7.2.11
    ports:
      - '1111:1111'
      - '8890:8890'
    shm_size: 4g
    environment:
      - DBA_PASSWORD=dba
```

### unixODBC 2.3.11
Installed using `sudo dnf install unixODBC-devel`

#### ~/.odbc.ini
```
; Data Source Name and associated Driver Section
VOS          = OpenLink Virtuoso ODBC Driver (Unicode)

; Data Source Name and associated Driver Library section
[VOS]
Description = Open Virtuoso
Driver      = /usr/local/virtuoso-opensource/lib/virtodbcu_r.so
Database    = Demo
Address     = localhost:1111
WideAsUTF16 = Yes
```

#### /etc/odbcinst.ini

Manually added the following lines to the end of the file:
```
# Driver from Virtuoso-SQLAlchemy
# Setup from the unixODBC pacakge
[VOS]
Description     = Open Virtuoso
Driver              = /usr/local/virtuoso-opensource/lib/virtodbcu_r.so
Database        = Demo
Address          = localhost:1111
WideAsUTF16     = Yes
```

#### virtodbcu_r.so
Copied from the virtuoso container to /usr/local/virtuoso-opensource/lib/ via `docker cp` (since we could not find it anywhere else):
```
sudo docker cp virtuoso_db_1:/opt/virtuoso-opensource/lib/virtodbcu_r.so /usr/local/virtuoso-opensource/lib/virtodbcu_r.so
```

**Comment by HughWilliams** [comment]
The DSN information should in `/etc/odbc.ini` rather than `/etc/odbcinst.ini`  for system DSNs  picked up by whatever user your application runs as ...

**Comment by eltonfss** [comment]
Hi @HughWilliams, thank you for your help!

I've tried the configuration you suggested and it seems to have worked, since now I'm getting a different error message:
```
 File "<virtualenvpath>/lib64/python3.11/site-packages/sqlalchemy/engine/default.py", line 616, in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sqlalchemy.exc.DBAPIError: (pyodbc.Error) ('01000', "[01000] [unixODBC][Driver Manager]Can't open lib '/usr/local/virtuoso-opensource/lib/virtodbcu_r.so' : file not found (0) (SQLDriverConnect)")
(Background on this error at: https://sqlalche.me/e/20/dbapi)
```

I've tried moving the `/virtuoso-opensource/lib/virtodbcu_r.so` to `<myhomefolder>`, and changing the .ini files to point to it.
Nonetheless, I got a similar error message:
```
 File "<virtualenvpath>/lib64/python3.11/site-packages/sqlalchemy/engine/default.py", line 616, in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sqlalchemy.exc.DBAPIError: (pyodbc.Error) ('01000', "[01000] [unixODBC][Driver Manager]Can't open lib '<myhomefolder>/virtuoso-opensource/lib/virtodbcu_r.so' : file not found (0) (SQLDriverConnect)")
(Background on this error at: https://sqlalche.me/e/20/dbapi)
```

Do you have any ideas about what else could I try to make this work?

**Comment by HughWilliams** [comment]
If the `virtodbcu_r.so` exist in either location, are you sure the file has read access as the user the application is running as ?

If you run the command `file virtodbc_r.so` does it return a file type indicating it is compatible/runnable on the OS you are running on ?

As the user the application is being run as, are you able to make a base unixODBC [isql](https://www.mankier.com/1/isql) connection , to prove the driver is usable, before attempting the connection via the application/python/pyodbc layers ?

**Comment by eltonfss** [comment]
Hi @HughWilliams, thanks again for your help!

To make sure the answer to your first question is yes, I've ran:
```sh
chmod 777 <myhomefolder>/virtuoso-opensource -R
```

Here is the output of the command `ls -lh` at `<myhomefolder>/virtuoso-opensource/lib/virtodbcu_r.so` after that:
```sh
total 3.9M
-rwxrwxrwx. 1 eltons eltons 3.9M Feb 12 07:43 virtodbcu_r.so
```

I tried running the test again after changing the permissions but I still getting the following error:
```
File "<myvirtualenv>/lib64/python3.11/site-packages/sqlalchemy/engine/default.py", line 616, in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sqlalchemy.exc.DBAPIError: (pyodbc.Error) ('01000', "[01000] [unixODBC][Driver Manager]Can't open lib '/home/eltons/virtuoso-opensource/lib/virtodbcu_r.so' : file not found (0) (SQLDriverConnect)")
```

With regards to your second question, here is the output of `file virtuoso_r.so`:
```
virtodbcu_r.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=ff316d2ce468a291dfb07f43181c45ff5133f0b0, with debug_info, not stripped
```

I've also tried running:
```sh
isql virtuoso+pyodbc://dba:dba@VOS -v
```
but received the following error with both the root user and my own user: 
```
[IM002][unixODBC][Driver Manager]Data source name not found and no default driver specified
[ISQL]ERROR: Could not SQLConnect
```

**Comment by HughWilliams** [comment]
Actually looking back at you original case notes you indicated having copied the `virtuoso_r.so` ODBC driver file from a Virtuoso **Ubuntu** docker container to your target **Fedora** container, which are not fully compatible, hence the `file not found` errors you are getting, which are not that `virtuoso_r.so` is not found but rather one of its dependant system libraries on Ubuntu that do not exist on Fedora.

We have a [Generic Linux Virtuoso open source](https://github.com/openlink/virtuoso-opensource/releases/download/v7.2.12/virtuoso-opensource.x86_64-generic_glibc25-linux-gnu.tar.gz) installer containing prebuilt generic binaries that will run on any glibc 2.5 or higher Linux distribution and it what you need to download and use the `virtuoso_r.so` ODBC driver file it contains

Also, have you actually read the unixODBC [isql](https://www.mankier.com/1/isql) documentation as the correct format for usage is `isql DSN [USER [PASSWORD]]` and not what you are passing ie `virtuoso+pyodbc://dba:dba@VOS -v` which is the higher level SQLAlchemy connect string format.

**Comment by eltonfss** [comment]
It worked! 🥳 🚀 

After replacing  the `virtodbcu_r.so` by the one in the [Generic Linux Virtuoso open source](https://github.com/openlink/virtuoso-opensource/releases/download/v7.2.12/virtuoso-opensource.x86_64-generic_glibc25-linux-gnu.tar.gz) installer I've been able to connect to Virtuoso running on docker and run create the tables and indexes for my tests.

I'll now start testing insert, select and update queries, but I believe the main problem of this issue has been solved.

@HughWilliams thank you very much for your help!
